File: Pharmaceutical_Company_Example/Structural/C++/Composite.cpp
#include <iostream>
#include <vector>
#include <string>

// Component
class RegulatoryProcess {
public:
  virtual void execute() = 0;
  virtual ~RegulatoryProcess() = default;
};

// Leaf
class Step : public RegulatoryProcess {
public:
  Step(const std::string& step_name) : step_name(step_name) {}
  
  void execute() override {
    std::cout << "Executing step: " << step_name << std::endl;
  }
  
private:
  std::string step_name;
};

// Composite
class CompositeProcess : public RegulatoryProcess {
public:
  CompositeProcess(const std::string& process_name) : process_name(process_name) {}
  
  void add(RegulatoryProcess* process) {
    children.push_back(process);
  }
  
  void remove(RegulatoryProcess* process) {
    children.erase(std::remove(children.begin(), children.end(), process), children.end());
  }
  
  void execute() override {
    std::cout << "Executing composite process: " << process_name << std::endl;
    for (auto& child : children) {
      child->execute();
    }
  }
  
private:
  std::string process_name;
  std::vector<RegulatoryProcess*> children;
};

int main() {
  // Leaf objects - individual steps
  Step clinical_trial_step("Clinical Trial Approval");
  Step safety_step("Safety Review");
  Step manufacturing_step("Manufacturing Validation");
  
  // Composite processes
  CompositeProcess clinical_process("Clinical Process");
  clinical_process.add(&clinical_trial_step);
  
  CompositeProcess regulatory_process("Regulatory Process");
  regulatory_process.add(&safety_step);
  regulatory_process.add(&manufacturing_step);
  
  CompositeProcess full_approval_process("Full Drug Approval Process");
  full_approval_process.add(&clinical_process);
  full_approval_process.add(&regulatory_process);
  
  // Execute the full approval process
  full_approval_process.execute();
  
  return 0;
}

File: Pharmaceutical_Company_Example/02_Structural/R/Composite.R

# Component
RegulatoryProcess <- setRefClass(
  "RegulatoryProcess",
  methods = list(
    execute = function() {
      stop("This method should be implemented by subclasses.")
    }
  )
)

# Leaf
Step <- setRefClass(
  "Step",
  contains = "RegulatoryProcess",
  fields = list(step_name = "character"),
  methods = list(
    initialize = function(step_name) {
      .self$step_name <- step_name
    },
    execute = function() {
      message("Executing step: ", .self$step_name)
    }
  )
)

# Composite
CompositeProcess <- setRefClass(
  "CompositeProcess",
  contains = "RegulatoryProcess",
  fields = list(process_name = "character", children = "list"),
  methods = list(
    initialize = function(process_name) {
      .self$process_name <- process_name
      .self$children <- list()
    },
    add = function(regulatory_process) {
      .self$children <- append(.self$children, list(regulatory_process))
    },
    remove = function(regulatory_process) {
      .self$children <- .self$children[.self$children != regulatory_process]
    },
    execute = function() {
      message("Executing composite process: ", .self$process_name)
      for (child in .self$children) {
        child$execute()
      }
    }
  )
)

# Leaf objects - individual steps
clinical_trial_step <- Step$new("Clinical Trial Approval")
safety_step <- Step$new("Safety Review")
manufacturing_step <- Step$new("Manufacturing Validation")

# Composite processes
clinical_process <- CompositeProcess$new("Clinical Process")
clinical_process$add(clinical_trial_step)

regulatory_process <- CompositeProcess$new("Regulatory Process")
regulatory_process$add(safety_step)
regulatory_process$add(manufacturing_step)

full_approval_process <- CompositeProcess$new("Full Drug Approval Process")
full_approval_process$add(clinical_process)
full_approval_process$add(regulatory_process)

full_approval_process$execute()
